Logical flow

samgdotson opened the pull request:
## Summary of changes
<!--- In one or more sentences, describe the PR you are submitting. -->
This PR adds a new dispatch model to `osier` called `LogicDispatchModel` that follows a hierarchy of rules to dispatch energy. It is much faster (~4x speedup) than the original `DispatchModel` but it is myopic and therefore results may be sub-optimal. Other benefits include:
- No solver required, commercial or otherwise. Thus `osier` is more self-contained.
- Easier to understand since the mathematics have been reduced to a sequence of comparisons.
- Drop-in replacement for `DispatchModel`. No change to syntax.
- Enables parallelization of the genetic algorithm driving the `CapacityExpansion` model.

>[!NOTE]
> There are a couple of new tests for new functionality. These tests should exceed the minimum code coverage required by the `pyproject.toml` file (75%). More tests are desired, but not necessary, here.

Edit: Closes #54 by changing `MW*hr` to `MWh`.

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] I have read the **CONTRIBUTING** document.
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

## Checklist for Reviewers

Reviewers should use [this link](https://arfc.github.io/manual/guides/pull_requests) to get to the
Review Checklist before they begin their review.

samgdotson committed:
adds comparison operators for technologies

samgdotson committed:
adds tests for tech comparison operators

samgdotson committed:
adds logical flow to storage tech

samgdotson committed:
merge conflicts

samgdotson committed:
Merge branch 'tech_library' of github.com:samgdotson/osier into logical_flow

samgdotson committed:
merge conf

samgdotson committed:
adds power behavior for technologies and tests

samgdotson committed:
updates the storage technology

samgdotson committed:
Merge branch 'tech_library' of github.com:samgdotson/osier into logical_flow

samgdotson committed:
fixes battery methods

samgdotson committed:
fixes battery

samgdotson committed:
adds tests for battery storage

samgdotson committed:
adds functions to utils and a general model class

samgdotson committed:
updates solve in logic model

samgdotson committed:
Merge branch 'debugging' of github.com:samgdotson/osier into logical_flow

samgdotson committed:
adds empty results formatting method

samgdotson committed:
minor changes

samgdotson committed:
merge conf

samgdotson committed:
Merge branch 'main' of github.com:arfc/osier into logical_flow

samgdotson committed:
updates init

samgdotson committed:
adds hierarchical dispatch example

samgdotson committed:
updates hierarchical dispatch notebook

samgdotson committed:
updates logic dispatch and example notebook

samgdotson committed:
updates logic flow and battery output

samgdotson committed:
adds test for logic model, issue with power_history attribute

samgdotson committed:
working logical flow model

samgdotson committed:
updates dispatch notebook

samgdotson committed:
adds parallelization tutorial

samgdotson committed:
updates dispatch tutorial with new model

samgdotson committed:
removes empty cell in notebook

samgdotson committed:
removes notebook

pep8speaks commented:
Hello @samgdotson! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:





* In the file [`osier/models/capacity_expansion.py`](https://github.com/arfc/osier/blob/5565b29c1cb281b4a6c2e7a498a3aaa874c239a9/osier/models/capacity_expansion.py):

> [Line 17:29](https://github.com/arfc/osier/blob/5565b29c1cb281b4a6c2e7a498a3aaa874c239a9/osier/models/capacity_expansion.py#L17): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ':'

* In the file [`osier/models/dispatch.py`](https://github.com/arfc/osier/blob/5565b29c1cb281b4a6c2e7a498a3aaa874c239a9/osier/models/dispatch.py):

> [Line 111:80](https://github.com/arfc/osier/blob/5565b29c1cb281b4a6c2e7a498a3aaa874c239a9/osier/models/dispatch.py#L111): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (92 > 79 characters)

* In the file [`osier/models/logic_dispatch.py`](https://github.com/arfc/osier/blob/5565b29c1cb281b4a6c2e7a498a3aaa874c239a9/osier/models/logic_dispatch.py):

> [Line 15:53](https://github.com/arfc/osier/blob/5565b29c1cb281b4a6c2e7a498a3aaa874c239a9/osier/models/logic_dispatch.py#L15): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 16:1](https://github.com/arfc/osier/blob/5565b29c1cb281b4a6c2e7a498a3aaa874c239a9/osier/models/logic_dispatch.py#L16): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 22:80](https://github.com/arfc/osier/blob/5565b29c1cb281b4a6c2e7a498a3aaa874c239a9/osier/models/logic_dispatch.py#L22): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (104 > 79 characters)
> [Line 45:80](https://github.com/arfc/osier/blob/5565b29c1cb281b4a6c2e7a498a3aaa874c239a9/osier/models/logic_dispatch.py#L45): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (89 > 79 characters)
> [Line 66:23](https://github.com/arfc/osier/blob/5565b29c1cb281b4a6c2e7a498a3aaa874c239a9/osier/models/logic_dispatch.py#L66): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 67:34](https://github.com/arfc/osier/blob/5565b29c1cb281b4a6c2e7a498a3aaa874c239a9/osier/models/logic_dispatch.py#L67): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 137:29](https://github.com/arfc/osier/blob/5565b29c1cb281b4a6c2e7a498a3aaa874c239a9/osier/models/logic_dispatch.py#L137): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 138:29](https://github.com/arfc/osier/blob/5565b29c1cb281b4a6c2e7a498a3aaa874c239a9/osier/models/logic_dispatch.py#L138): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 146:51](https://github.com/arfc/osier/blob/5565b29c1cb281b4a6c2e7a498a3aaa874c239a9/osier/models/logic_dispatch.py#L146): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace





* In the file [`osier/technology.py`](https://github.com/arfc/osier/blob/5565b29c1cb281b4a6c2e7a498a3aaa874c239a9/osier/technology.py):

> [Line 634:80](https://github.com/arfc/osier/blob/5565b29c1cb281b4a6c2e7a498a3aaa874c239a9/osier/technology.py#L634): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (86 > 79 characters)
> [Line 636:1](https://github.com/arfc/osier/blob/5565b29c1cb281b4a6c2e7a498a3aaa874c239a9/osier/technology.py#L636): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 762:80](https://github.com/arfc/osier/blob/5565b29c1cb281b4a6c2e7a498a3aaa874c239a9/osier/technology.py#L762): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (86 > 79 characters)
> [Line 953:1](https://github.com/arfc/osier/blob/5565b29c1cb281b4a6c2e7a498a3aaa874c239a9/osier/technology.py#L953): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 960:1](https://github.com/arfc/osier/blob/5565b29c1cb281b4a6c2e7a498a3aaa874c239a9/osier/technology.py#L960): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace









##### Comment last updated at 2025-01-15 20:22:32 UTC

samgdotson committed:
autopep8

samgdotson committed:
autopep8 again

samgdotson committed:
manual changes

LukeSeifert commented on osier/models/logic_dispatch.py:
```diff
@@ -0,0 +1,100 @@
+                data[f"{t.technology_name}_charge"] = unyt_array(
+                    t.charge_history).to_ndarray()
+        data["Curtailment"] = np.array(
+            [v if v <= 0 else 0 for v in self.covered_demand])
```
I understand using `t` for `technology`, but what is `v` here?

samgdotson replied:
v is for voltage.

LukeSeifert commented on osier/models/logic_dispatch.py:
```diff
@@ -0,0 +1,100 @@
+    def solve(self):
+        """
+        This function executes the model solve with a rule-based approach.
+        Net demand is copied, then the technology histories are reset.
```
```suggestion
        Net demand is copied, then the technology histories are reset.
        This function executes the model solve with a rule-based approach.
```
The code appears to be in this order

LukeSeifert commented on osier/models/logic_dispatch.py:
```diff
@@ -0,0 +1,100 @@
+LARGE_NUMBER = 1e20
+
+
+class LogicDispatchModel(OsierModel):
```
No class doc string?

yardasol replied:
Docstring?

LukeSeifert commented on osier/tech_library.py:
```diff
@@ -96,7 +96,7 @@
                             storage_duration=4,
                             efficiency=0.85,
-                            initial_storage=0,
+                            initial_storage=0.0*MW*hr,
```
Why not use `MWh`?

yardasol replied:
Should this be `MWh` to reflect the shift from `MW*hr` to `MWh`?

LukeSeifert commented on osier/technology.py:
```diff
@@ -306,17 +306,58 @@ def __init__(self,
+        """Tests greater or equal to."""
+        if (self.variable_cost == tech.variable_cost):
+            return self.efficiency >= tech.efficiency
+        elif self.variable_cost >= tech.variable_cost:
+            return True
+        else:
+            return False
```
```suggestion
        else:
            return self.variable_cost >= tech.variable_cost
```

samgdotson replied:
Could you elaborate on this?

LukeSeifert commented on osier/technology.py:
```diff
@@ -306,17 +306,58 @@ def __init__(self,
+        elif self.variable_cost <= tech.variable_cost:
+            return True
+        else:
+            return False
```
```suggestion
        else:
            return self.variable_cost <= tech.variable_cost
```

LukeSeifert commented on osier/technology.py:
```diff
@@ -306,17 +306,58 @@ def __init__(self,
+        elif self.variable_cost < tech.variable_cost:
+            return True
+        else:
+            return False
```
```suggestion
        else:
            return self.variable_cost < tech.variable_cost
```

LukeSeifert commented on osier/technology.py:
```diff
@@ -306,17 +306,58 @@ def __init__(self,
+
+        if (self.variable_cost == tech.variable_cost):
+            return self.efficiency > tech.efficiency
+        elif self.variable_cost > tech.variable_cost:
             return True
         else:
             return False
```
```suggestion
        else:
            return self.variable_cost > tech.variable_cost
```

LukeSeifert commented on osier/technology.py:
```diff
@@ -564,21 +608,44 @@ def to_dataframe(self, cast_to_string=True):
+        ----------
+        demand : :class:`unyt.unyt_quantity`
+            The demand at a particular timestep. Must be a :class:`unyt.unyt_quantity`
+            to avoid ambiguity.
```
Missing `Returns` in docstring

LukeSeifert commented on osier/technology.py:
```diff
@@ -645,6 +712,69 @@ def ramp_down(self):
+        Parameters
+        ----------
+        time_delta : :class:`unyt.unyt_quantity`
+            The difference between two timesteps. Default is one hour.
```
Missing `Returns` in docstring (same for next couple functions)

LukeSeifert commented on osier/technology.py:
```diff
@@ -724,3 +858,58 @@ def initial_storage(self, value):
+        self.power_level = self.capacity
+        self.charge_history = []
+
+    def discharge(self, demand: unyt_quantity, time_delta=1 * hr):
```
These functions are missing docstrings

yardasol replied:
Docstring?

yardasol commented on osier/models/capacity_expansion.py:
```diff
@@ -7,12 +7,17 @@
 
 LARGE_NUMBER = 1e20
 
+dispatch_models = {'lp': DispatchModel,
+                   'hierarchical': LogicDispatchModel,
+                   'logical': LogicDispatchModel,
+                   'rule_based': LogicDispatchModel}
```
`lp` feels a bit too brief. Consider making it more descriptive. And why are there 3 keys for selecting `LogicDispatchModel`? Consider condensing those into one key.

LukeSeifert commented on tests/test_technology.py:
```diff
@@ -3,10 +3,11 @@
+from osier.tech_library import *
 
 TECH_NAME = "PlanetExpress"
 energy_unyt = 10.0 * MW * hr
```
You import `MWh`, consider using here

LukeSeifert commented on tests/test_technology.py:
```diff
@@ -239,7 +254,7 @@ def test_om_cost_variable(advanced_tech):
         advanced_tech.om_cost_variable = spec_energy_str
     assert advanced_tech.om_cost_variable.value == 0.0
-    assert advanced_tech.om_cost_variable.units == (MW*hr)**-1
+    assert advanced_tech.om_cost_variable.units == (MW * hr)**-1
```
Consider using `MWh`

LukeSeifert commented on tests/test_technology.py:
```diff
@@ -296,7 +311,7 @@ def test_fuel_cost(advanced_tech):
 
 def test_co2_rate(advanced_tech):
-    expected_unit = megatonnes*(MW*hr)**-1
+    expected_unit = megatonnes * (MW * hr)**-1
```
Consider using MWh

LukeSeifert reviewed (commented):
Looks good! I just had a couple comments on some docstrings and the `MWh` usage. I also had some trouble running the tests myself. The tests fail with installation method 1 (from source) and method 2 for me after I cloned your [repo](https://github.com/samgdotson/osier/tree/logical_flow) and switched to the `logical_flow` branch. Do you know what I need to do to test this version of `osier`?

yardasol commented on osier/models/model.py:
```diff
@@ -0,0 +1,80 @@
+              'T': 'minute'}
+
+
+class OsierModel():
```
This class has a lot of shared code with `DispatchModel`. Will `DispatchModel` be refactored to inherit from `OsierModel`?

samgdotson replied:
Yes, that is the dream. I realized there was going to be significant overlap in these two classes but not enough overlap to warrant directly inheriting from `DispatchModel`.

yardasol commented on osier/technology.py:
```diff
@@ -306,17 +306,58 @@ def __init__(self,
+            return False
+
+    def __le__(self, tech) -> bool:
+        """Tests greater or equal to."""
```
```suggestion
        """Tests less or equal to."""
```

yardasol commented on osier/technology.py:
```diff
@@ -306,17 +306,58 @@ def __init__(self,
+            return False
+
+    def __lt__(self, tech) -> bool:
+        """Tests greater or equal to."""
```
```suggestion
        """Tests less to."""
```

yardasol commented on osier/technology.py:
```diff
@@ -306,17 +306,58 @@ def __init__(self,
+            return False
+
+    def __gt__(self, tech) -> bool:
+        """Tests greater or equal to."""
```
```suggestion
        """Tests greater to."""
```

yardasol commented on osier/utils.py:
```diff
@@ -84,7 +83,7 @@ def get_dispatchable_techs(technology_list):
     Returns
     -------
-    tech_names : list of :class:`osier.Technology`
+    dispatchable_techs : list of :class:`osier.Technology`
```
Good name change

yardasol commented on osier/utils.py:
```diff
@@ -114,6 +113,50 @@ def get_nondispatchable_techs(technology_list):
+
+    Returns
+    -------
+    storage_techs : list of :class:`osier.Technology`
```
```suggestion
    nonstorage_techs : list of :class:`osier.Technology`
```

yardasol commented on osier/technology.py:
```diff
@@ -724,3 +858,58 @@ def initial_storage(self, value):
+        self.charge_history.append(0 * demand.units)
+        return self.power_level.to(demand.units)
+
+    def charge(self, surplus, time_delta=1 * hr):
```
Docstring?

yardasol reviewed (changes requested):
Hi @samgdotson, great work on this PR. A couple items

- Docpages have not been updated to hold the new classes
- Line-items (see comments)

Additionally, it's not clear to me where the no-solver change is, nor where parallelization of the genetic algorithm is enabled. If you could describe these changes in a bit more detail that'd be wonderful!

Commit:
adds docstring to logic model

samgdotson committed:
adds logic model to docs

samgdotson committed:
merge conflict

samgdotson reviewed (commented):
Thanks for the reviews @yardasol and @LukeSeifert!

samgdotson committed:
changes from code review

samgdotson committed:
fixes doc string

samgdotson commented:
@yardasol, just addressing your comments/questions:

1. There is no parallelization algorithm in this PR. The PR _enables_ parallelization. 
2. The "no-solver" change lives in `CapacityExpansion` where I added a flag called `model_engine`.

The docs should be updated, now. Also, I'm not sure why the GH actions are failing, but it's not anything related to the code.

LukeSeifert commented on osier/technology.py:
```diff
@@ -724,3 +870,94 @@ def initial_storage(self, value):
+
+    def power_output(self, v, time_delta=1 * hr):
+        """
+        Calculates the power output given a demand value.
```
Missing parameters

LukeSeifert reviewed (approved):
Looks good! I made a comment on one docstring that has the parameters missing, everything else seems good to me

samgdotson committed:
adds parameters doc string

samgdotson commented:
@LukeSeifert @yardasol if you're satisfied with the changes feel free to merge (in spite of the failing checks -- I think it is a GH issue).

samgdotson committed:
updates mamba setup step in gh action

samgdotson commented:
Okay, I think the issue is because Mambaforge is no longer being maintained, so the installer is looking for a link that doesn't go anywhere, anymore. See this [release](https://github.com/conda-forge/miniforge/releases/tag/24.11.2-0).

Working on a fix.

samgdotson committed:
removes

samgdotson commented:
@yardasol, checks pass, now. Are you satisfied with the changes?

@LukeSeifert, regarding testing locally, I can't say for sure what's wrong because I haven't seen the error message. However, I speculate that the errors might be because you haven't "installed" the branch. Try creating an editable installation (as described in the [contributing guide](https://osier.readthedocs.io/en/latest/contrib.html)) with 

```bash
python -m pip install -e .[doc]
```

This way, you can switch between branches at will and the "installed version" of `osier` will follow code changes (including changing branches).

yardasol commented on osier/models/capacity_expansion.py:
```diff
@@ -72,6 +77,11 @@ class CapacityExpansion(ElementwiseProblem):
         Indicates which solver to use. May require separate installation.
         Accepts: ['cplex', 'cbc', 'glpk']. Other solvers will be added in the
         future.
+    model_engine : str
+        Determines which dispatch algorithm to use.
+        Accepts: ['optimal', 'logical'] where 'optimal' will use a linear
+        program and 'logical' will use a myopic rule-based approach.
+        Default is 'optimal'.
```
Good description.

yardasol commented on osier/models/capacity_expansion.py:
```diff
@@ -7,12 +7,17 @@
 
 LARGE_NUMBER = 1e20
 
+dispatch_models = {
+                   'optimal':DispatchModel,
+                   'logical': LogicDispatchModel
+                   }
```
Good!

yardasol commented on osier/technology.py:
```diff
@@ -645,6 +709,84 @@ def ramp_down(self):
+
+        Returns
+        -------
+        max_power : :class:`unyt.unyt_quantity`
+            The maximum achievable power level.
```
```suggestion
        min_power : :class:`unyt.unyt_quantity`
            The minimum achievable power level.
```

yardasol reviewed (approved):
Good changed @samgdotson. I found one little snag but other than that we're good to go!

samgdotson committed:
Update osier/technology.py

Co-authored-by: Olek <45364492+yardasol@users.noreply.github.com>

samgdotson commented:
Thanks @yardasol! Feel free to merge whenever!
